File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int n = nums.length;
        if(nums[0]==target) return 0;
        if(nums[n-1]==target) return n-1;
        int i=0, j=n-1;
        while(i <= j){
            int mid = i+(j-i)/2;
            if(nums[mid]==target){
                return mid;
            }
            if(nums[i] <= nums[mid]){
                if(nums[i] <= target && nums[mid]>=target){
                    j=mid-1;
                }else{
                    i= mid+1;
                }
            }else{
                if(nums[mid] <=target && nums[j] >= target){
                    i=mid+1;
                }else{
                    j=mid-1;
                }
            }
        }
        return -1;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int low=0, high=nums.size()-1;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(nums[mid]==target){
                return mid;
            }
            //left portion search
            if(nums[low] <= nums[mid]){
                if(target > nums[mid] || target < nums[low]){
                    low=mid+1;
                }else{
                    high=mid-1;
                }
            }
            //right portion
            else{
                if(target < nums[mid] || target >nums[high]){
                    high=mid-1;
                }else{
                    low=mid+1;
                }
            }
        }
        return -1;
    }
};